### ISSUE TITLE
Predict_class.py - cannot find models

### ISSUE THREAD

**Issue by MattLomb** [dibollinger/CookieBlock-Consent-Classifier#3]
Hi! 
I'm doing a research over GDPR compliance for Italian websites and I've found your beautiful work (great job!).

I've found in the documentation the instruction to run the **predict_class.py** script and I'm trying to run it, unfortunately without success because I'm not finding the **model** to pass as input to the script.
I've tried with the models present in the /classifiers folder, but no one works.

I leave here an example of error:
```
python3 predict_class.py ./classifiers/xgboost_sparse_to_libsvm.py ../test/json_test.json 
Unrecognized model type for './classifiers/xgboost_sparse_to_libsvm.py'
```
(_../test/json_test.json_ is the JSON data I've created to test the script, formatted as you suggested)

Can you please help me? Where can I find the models?
Or better, there are pre-trained models or do I need to train them using the one in the classifiers section?
(I've formatted the JSON object to give in input as you've described in the documentation section).

Thank you in advance,
Matteo.
@dibollinger @Bender250

**Comment by Bender250** [comment]
Hi Matteo!

Pretrained models together with training data are available at [Zenodo](https://zenodo.org/record/5838646). The README in that record should guide you in using it. Some additional information is also in the [artifacts appendix of the paper](https://karelkubicek.github.io/assets/pdf/Automating_Cookie_Consent_and_GDPR_Violation_Detection.pdf).

The model expect feature extraction according to feature processing in this repo. If you want even a simpler way of collecting the classification, you can just simply download CookieBlock, allow *Keep track of cookie history* in settings, *Pause cookie removal* (to not interfere with websites functionality), and browser the selected websites. Then you can find predictions under button *Open cookie configuration* with predictions or *View full cookie history JSON* which even shows the changes to cookies.

Let me know if I can help with something else. And once you have the paper written, would you share it to me by email <karel.kubicek@inf.ethz.ch>? I am curious to see your findings!

**Comment by MattLomb** [comment]
Hi Karel!
Thank you in advance for the response.

I am already working with the browser extension and it is working fine.
I would like to use the script predict_class.py to classify the JSON cookie passed as argument and to automatize my study. 
So, I've extracted the pretrained xgb model in the dataset hosted in Zenodo and I'm currently making some experiments passing as input to the scripts some JSON I've extracted from some Italian websites that contains a lot of GDPR violations.
I will update you on the evolution of the project and on the achievements.

Thank you for the support! 

:smile:

**Comment by MattLomb** [comment]
I'm running the predict_class.py script, but it doesn't give me any class in output and shows also some errors :cry: 
I've runned the script using the pretrained xgb model and passing as input the test_case.json found in the training_data folder.
I leave here the result of the execution:
```
python3 predict_class.py ../xgbmodel_full_20210517_204300.xgb ./training_data/test_case.json 
Extracting Features...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 674.11it/s]
Predicting Labels...
Confusion Matrix:
[[1 0 0 0]
 [0 0 0 0]
 [0 0 0 0]
 [0 1 0 0]]
Total Accuracy Count: 1
Total Accuracy Ratio: 0.5
Micro Precision: 0.5
Micro Recall: 0.5
Micro F1Score: 0.5
/home/matteo/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/matteo/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
Macro Precision: 0.3333333333333333
Macro Recall: 0.3333333333333333
Macro F1Score: 0.3333333333333333
/home/matteo/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/matteo/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
Weighted Precision: 0.5
Weighted Recall: 0.5
Weighted F1Score: 0.5
/home/matteo/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/matteo/.local/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
Precision for each class: [1. 0. 0.]
Recall for each class: [1. 0. 0.]
F1Score for each class: [1. 0. 0.]
-------------------------------
/home/matteo/Desktop/CookieBlock-Consent-Classifier/classifiers/utils.py:233: RuntimeWarning: invalid value encountered in long_scalars
  recall = confusion_matrix[i, i] / np.sum(confusion_matrix[i, :])
/home/matteo/Desktop/CookieBlock-Consent-Classifier/classifiers/utils.py:232: RuntimeWarning: invalid value encountered in long_scalars
  precision = confusion_matrix[i, i] / np.sum(confusion_matrix[:, i])
(Old Method) Total Accuracy: 50.000%
(Old Method) Precision: [ 1.  0. nan nan]
(Old Method) Recall: [ 1. nan nan  0.]
(Old Method) F1 Scores: [ 1. nan nan nan]
Predictions on Consent Cookies: {'OptanonConsent': [0, 0, 0, 0], 'CookieConsent': [0, 0, 0, 0]}
```

Can you please help me?
@Bender250

**Comment by Bender250** [comment]
This is caused by some classes being empty (no cookie classified to this class). Once you start having inputs with all classes, these warnings will disappear. It should not matter for what you do.

I will take a note of this problem and hopefully in future fix the code to suppress these warnings.

**Comment by MattLomb** [comment]
I'm doing some tests using a Google cookie, used for analytics, but unfortunately the script doesn't give me any classifier.
`{
    "3": {
      "name": "_ga_S0X9SBF32B",
      "domain": ".hsr.it",
      "path": "/",
      "first_party_domain": "",
      "label": 1,
      "cmp_origin": 0,
       "variable_data": [
        {
        "value": "GS1.1.1678553657.1.0.1678553657.0.0.0",
        "expiry": 1741625657,
        "session": "true",
        "http_only": false,
        "host_only": true,
        "secure": false,
        "same_site": "lax"
        }
      ]
    }
   }`
Moreover, passing as "same_site" value "no restriction", the script fails. 

Am I doing something wrong? Or there are problems in the script?

Waiting for your response, thanks in advance.
@Bender250 @dibollinger

**Comment by MattLomb** [comment]
Hi guys, can you please help me?

**Comment by dibollinger** [comment]
> I'm running the predict_class.py script, but it doesn't give me any class in output and shows also some errors

It did give you the predictions, they are visible in the confusion matrix. One cookie which was labelled as necessary received the necessary class, and one which was labelled as "Advertising" received the "Functionality" prediction. Note however that this script was intended to test the model with a large number of cookies, and output statistics, and was not intended to output individual labels per cookie.

> Moreover, passing as "same_site" value "no restriction", the script fails.

This needs to be formatted as "no_restriction" with an underscore, as this is one of three formats produced by the OpenWPM crawler.

> I'm doing some tests using a Google cookie, used for analytics, but unfortunately the script doesn't give me any classifier.

To train a classifier you need a large number of cookies, and you'll need to run the feature extractor on them first before they can be transformed into a model. It won't work with a single cookie.

Closing this as the main question has been answered.
